bip-csfs.mediawiki

```diff
@@ -0,0 +1,124 @@
+
+==Summary==
+
+We propose replacing OP_NOP5 (0xb4) in bitcoin script with
+<code>OP_CHECKSIGFROMSTACKVERIFY</code>. When verifying taproot script spends having
+leaf version 0xc0 (as defined in BIP342), we propose <code>OP_CHECKSIGFROMSTACK</code>
+to replace <code>OP_SUCCESS204</code> (0xcc).
+
+<code>OP_CHECKSIGFROMSTACK</code> and <code>OP_CHECKSIGFROMSTACKVERIFY</code>
+have semantics similar to <code>OP_CHECKSIG</code> and
+<code>OP_CHECKSIGVERIFY</code> respectively, as specified below.
+
+For legacy and segwit v0 scripts, 32-byte and 33-byte keys are supported for
+BIP340 signature verification and ECDSA signature verification respectively.
+
+==Specification==
+
+* If fewer than 3 elements are on the stack, the script MUST fail and terminate immediately.
+* The public key (top element), message (second to top element), and signature (third from top element) are read from the stack.
+* For <code>OP_CHECKSIGFROMSTACK</code> the top three elements are popped from the stack.
+* If the public key size is zero, the script MUST fail and terminate immediately.
+* If the public key size is 32 bytes, it is considered to be a public key as described in BIP340:
+** If the signature is not the empty vector, the signature is validated against the public key and message according to BIP340. Validation failure in this case immediately terminates script execution with failure.
+* For legacy and segwit v0 scripts, if the public key size is 33 bytes and its first byte is 0x02 or 0x03, it is considered a compressed public key as described in BIP137:
+** If the signature is not the empty vector, the signature is validated against the public key and message using ECDSA. Validation failure in this case immediately terminates script execution with failure.
+* If the public key size is not zero, and it is not a BIP340 public key, nor a BIP137 compressed public key; the public key is of an unknown public key type, and no actual signature verification is applied. During script execution of signature opcodes they behave exactly as known public key types except that signature validation is considered to be successful.
+* If the script did not fail and terminate before this step, regardless of the public key type:
+** If the signature is the empty vector:
+*** For <code>OP_CHECKSIGFROMSTACKVERIFY</code>, the script MUST fail and terminate immediately.
+*** For <code>OP_CHECKSIGFROMSTACK</code>, an empty vector is pushed onto the stack, and execution continues with the next opcode.
+** If the signature is not the empty vector:
+*** For tapscript 0xc0, the opcode is counted towards the sigops budget as described in BIP341.
+*** For legacy and segwit v0, the opcode is counted towards the sigops limit, as described in BIP141
+*** For <code>OP_CHECKSIGFROMSTACKVERIFY</code>, execution continues without any further changes to the stack.
+*** For <code>OP_CHECKSIGFROMSTACK</code>, a 1-byte value 0x01 is pushed onto the stack.
```

Comment:
I wonder if there's a use for a `OP_CHECKSIGFROMSTACKADD` variant...

So something that would be compatible with usage of `OP_CHECKSIGADD`. On the other hand, it is possible to replicate such behavior with `OP_CHECKSIGFROMSTACK` in combination with other existing op codes, eg: 
```
OP_TOALTSTACK (assuming the number N is on the top of the stack)
<pubkey> OP_CHECKSIGFROMSTACK
OP_IF
    OP_DROP (drop the 0x01 pushed in the success case)
    OP_FROMALTSACK (bring back the accumulator) 
    1 OP_ADD (acc += 1)
OP_ELSE
    OP_DROP (drop the empty vector) 
    OP_FROMALTSACK (put unchanged accumulator back on the stack)
OP_ENDIF
```

Reply:
Yep. I had the same internal conversation and concluded that the rare cases where CSFSA behavior is needed using the altstack is an acceptable alternative.

Compared to standard CHECKSIG where CHECKMULTISIG was in active use on legacy/segwitv0 wallets, new wallets or protocols built on CSFS are much more likely to use MuSig2 or FROST when multiple signers are insolved, or (as BitGo's 2-of-3 MuSig2/tapscript wallet does) to have the different signing combinations split between tapscript paths where CSFSV+CSFS are more efficient than CSFSA.

Reply:
Keeping this BIP to 1 opcode. Other CSFS(A/V) variants can be added later if they are specifically needed.